Add design details and principles (#386)
repository-service-tuf/repository-service-tuf: docs/source/devel/design.rst

###################
Architecture Design
###################

The principles
##############

   * RSTUF uses `The Update Framework (TUF) <http://www.theupdateframework.io>`_.

     - It enables RSTUF to be artifact agnostic.

   * RSTUF is easy to deploy.

   * RSTUF is API first designed.

>      - RSTUF is language agnostic, making any language use/integrate easily.

Author: I rephrased it, using part of your suggestion but tried to clarify all the context.

   * RSTUF is process agnostic.

>      - Adding/removing artifacts goes with the existing releasing/publishing
>        process.

Reviewer: ```suggestion
     - The addition or removal of artifacts doesn't interfere 
     with the existing release/publish processes.
```

   * RSTUF focuses on scalability and Metadata consistency.

   * RSTUF guides users on TUF processes.

The architecture design principles:
###################################

The Repository Service for TUF (RSTUF) has two services
(``repository-service-tuf-api``, ``repository-service-tuf-worker``) and one
command line tool (``repository-service-cli``), as described in
:ref:`guide/general/introduction:RSTUF Components`.

.. note::
    Other services can be added as optional or required

RSTUF also requires some third-party services described in
:ref:`guide/deployment/planning/deployment:Required Infrastructure Services`.

.. image:: /_static/2_1_rstuf.png

The below definitions allow RSTUF API and RSTUF Worker scalability.

RSTUF is Asynchronous
=====================

    * Every API request is a task.

>     * RSTUF centralizes all tasks in the Message :ref:`devel/design:Message Queue`.

Reviewer: ```suggestion
    * RSTUF centralizes all tasks in the :ref:`devel/design:Message Queue`.
```

    * RSTUF stores all task results in the :ref:`devel/design:Backend Result`.
    * :ref:`devel/design:Repository Service TUF API` **publishes** RSTUF tasks.
    * :ref:`devel/design:Repository Service TUF Worker` **consumes** RSTUF tasks.

RSTUF Repository Settings/Configuration
=======================================

    * :ref:`guide/general/introduction:RSTUF Components` are RSTUF
      **stateless** about the Repository Settings, relying on

>       :ref:`devel/design:RSTUF settings`.

Reviewer: That one is a little strangely formulated.
Usually when you say `stateless` you say that something is overall stateless and not just for a particular setting.
Also, it's a little unclear what is `devel/design:RSTUF settings`.

Author: I gave more context about the stateless and fixed the reference, adding the `TUF Repository Settings` as a sub-title of `RSTUF Repository Settings/Configuration`

    * :ref:`devel/design:RSTUF settings` stores in
      :ref:`devel/design:Redis` using `Dynaconf <https://www.dynaconf.com>`_.

    * :ref:`devel/design:Repository Service TUF Worker` **writes** settings
    * :ref:`devel/design:Repository Service TUF API` **reads** settings.

        .. note::
          A unique exception is when a bootstrap process fails and the API

>           **writes** it back to ``None``.

Reviewer: This note looks a little bit out of place:
<img width="387" alt="image" src="https://github.com/repository-service-tuf/repository-service-tuf/assets/16246778/db80e476-3cf6-4dfc-9aad-a81062bc0c6b">

Is there a better place for this?

Author: The place is correct. I will reformulate the sentence.

Author: I rephrased the sentence and gave the full context of the exception.

    .. list-table:: RSTUF reserved settings/configuration
        :header-rows: 1
        :widths: 30 30 40

        * - Key
          - Value(s)
          - Description
        * - ``BOOTSTRAP``
          - | ``None``
            | ``<task id>``
            | ``pre-<task id>``
            | ``sign-<task id>``
          - | RSTUF bootstrap state
            | ``None``: No bootstrap
            | ``<task id>``: Finished
            | ``pre-<task id>``: Initial process
            | ``sign-<task id>``: Signing process
        * - ``<ROLE NAME>_EXPIRATION``
          - | ``int``

>             | days
>           - | Role Metadata expiration setting

Reviewer: I don't see why you explain the values inside the `values` column when there is a specific `description` column for that.
I mean these four settings precisly:
<img width="353" alt="image" src="https://github.com/repository-service-tuf/repository-service-tuf/assets/16246778/c1e59f28-c4ee-4a48-91ea-b7ee9ecdacce">

I have suggested below how to fix them:
```suggestion
          - | Role Metadata expiration policy in days
```

            | It uses the role name uppercase
            | Example: ``ROOT_EXPIRATION``
        * - ``<ROLE NAME>_NUM_KEYS``
          - | ``int``

>             | number of keys
>           - | Role number of keys

Reviewer: ```suggestion
          - | Role number of keys
```

            | It uses the role name uppercase
            | Example: ``ROOT_NUM_KEYS``
        * - ``<ROLE NAME>_THRESHOLD``
          - | ``int``

>             | keys threshold
>           - | Role threshold

Reviewer: ```suggestion
          - | Role key threshold
```

            | It uses the role name uppercase
            | Example: ``ROOT_THRESHOLD``
        * - ``NUMBER_OF_DELEGATED_BINS``
          - | ``int``
            | Number of delegated bins
          - number of delegated hash bin roles
        * - ``SIGNING_<ROLE NAME>``
          - | ``None``
            | ``<json>``
          - | ``None``: No pending signature(s)
            | ``json``: TUF Metadata pending signature
            | It uses the role name uppercase
            | Example ``SIGNING_ROOT``

Target Files and Target Roles
=============================

>     * The TUF top Role Targets is a role only for delegation. This role does
>       not register target files (artifacts).

Reviewer: ```suggestion    
    * The TUF top-level Targets Role is only used for delegation. 
This role does not register target files (artifacts).
```

      - Writing the TUF Metadata in the Storage Service  is limited to
        :ref:`devel/design:Repository Service TUF Worker`
      - The Storage Service is the only public data

RSTUF Components Design
#######################

Repository Service TUF API
==========================

* Integration (add/remove artifacts)
* TUF metadata process (bootstrap, sign, update, etc)

.. note::
    The service can implement other features without interfering with the
    RSTUF principles and architecture design principles.

`See component development documentation
<https://repository-service-tuf.readthedocs.io/projects/rstuf-api/en/latest/devel/>`_.

Repository Service TUF Worker
=============================

* Manages the TUF metadata
* Manages the Key Vault and Storage Services

.. note::
    The service can implement other features without interfering with the
    RSTUF principles and architecture design principles.

`See component development documentation
<https://repository-service-tuf.readthedocs.io/projects/rstuf-worker/en/latest/devel/>`_.

Repository Service TUF CLI
==========================

* Command Line Interface for the API

> * Guide users in the process

Reviewer: ```suggestion
* Guide users in the processes
```

.. note::
    The service can implement other features without interfering with the
    RSTUF principles and architecture design principles.

`See component development documentation
<https://repository-service-tuf.readthedocs.io/projects/rstuf-cli/en/latest/devel/>`_.

RSTUF Infrastructure Services Design
####################################

The Infrastructure Services have key functionality on RSTUF

Message Queue
=============

* It is a centralized queue

Backend Result
==============

> * It is a centralized backend result for tasks

Reviewer: Is this a service or what is it?
The description doesn't give much information.

Author: I will add more information

Author: I added more the context for Celery, included the same to message queue

###################
Architecture Design
###################

The principles
##############

   * RSTUF uses `The Update Framework (TUF) <http://www.theupdateframework.io>`_.

     - It enables RSTUF to be artifact agnostic.

   * RSTUF is easy to deploy.

   * RSTUF has API-first design.

     - RSTUF is language agnostic, allowing any programming language integration

   * RSTUF is process agnostic.

     - Add/Remove artifacts doesn't interfere with the existing organization

>        release/publish processes.

Author: It is still strange for me, maybe:
`Add/Remove artifacts to TUF Metadata doesn't interfere with the existing release/publish artifact processes.`

    Other services can be added as optional or required

RSTUF also requires some third-party services described in
:ref:`guide/deployment/planning/deployment:Required Infrastructure Services`.

.. image:: /_static/2_1_rstuf.png

The below definitions allow RSTUF API and RSTUF Worker scalability.

RSTUF is Asynchronous
=====================

    * RSTUF uses `Celery <https://docs.celeryq.dev>`_.
    * Every API request is a task.
    * RSTUF centralizes all tasks in the :ref:`devel/design:Message Queue`.
    * RSTUF stores all task results in the :ref:`devel/design:Backend Result`.
    * :ref:`devel/design:Repository Service TUF API` **publishes** RSTUF tasks.
    * :ref:`devel/design:Repository Service TUF Worker` **consumes** RSTUF tasks.

RSTUF Repository Settings/Configuration
=======================================

    * :ref:`guide/general/introduction:RSTUF Components` are **stateless**

      - Components are configured by runtime environment variables
      - Components uses runtime :ref:`devel/design:TUF Repository Settings`

    * :ref:`devel/design:Redis` stores the
      :ref:`devel/design:TUF Repository Settings` using
      `Dynaconf <https://www.dynaconf.com>`_.

>     * :ref:`devel/design:Repository Service TUF Worker` **writes** settings

Reviewer: ```suggestion
    * :ref:`devel/design:Repository Service TUF Worker` **writes** settings.
```

    * :ref:`devel/design:Repository Service TUF API` **reads** settings.

        .. note::
          A single exception is during a bootstrap process. If the
          :ref:`devel/design:Repository Service TUF API` detects a failure
          **writes** :ref:`devel/design:TUF Repository Settings`
          ``BOOTSTRAP`` to ``None``.

TUF Repository Settings
-----------------------

TUF Repository Setting are key configuration for the Metadata Repository
operations.

>     .. list-table:: RSTUF reserved settings/configuration

Author: I didn't mention it because there is no use case for this. 
It can be removed in the future.

Repository Service TUF CLI
==========================

* Command Line Interface for the API
* Guide users in the processes

.. note::
    The service can implement other features without interfering with the
    RSTUF principles and architecture design principles.

`See component development documentation
<https://repository-service-tuf.readthedocs.io/projects/rstuf-cli/en/latest/devel/>`_.

RSTUF Infrastructure Services Design
####################################

The Infrastructure Services have key functionality on RSTUF

Message Queue
=============

* It is a centralized queue service for tasks
* This queue is used as `Broker by Celery
  <https://docs.celeryq.dev/en/stable/getting-started/backends-and-brokers/index.html#broker-overview>`_.
* :ref:`devel/design:Repository Service TUF Worker` defines the supported Queue
  servers.

Backend Result
==============

* It is a centralized `backend result used by Celery for task results
  <https://docs.celeryq.dev/en/stable/getting-started/backends-and-brokers/index.html>`_
* :ref:`devel/design:Repository Service TUF Worker` defines the supported
  backend results servers.

Redis
=====

> * It is a centralized key/cache service

Reviewer: ```suggestion
* It is a centralized key/cache service.
```

* Stores :ref:`devel/design:RSTUF Repository Settings/Configuration`
* Optional:

>   - Used as :ref:`devel/design:Message Queue`
>   - Used as :ref:`devel/design:Backend Result`

Reviewer: ```suggestion
  - Used as :ref:`devel/design:Message Queue`.
  - Used as :ref:`devel/design:Backend Result`.
```
